File: packages/sdk-python/netcrawl/nodes.py
"""
netcrawl/nodes.py

Typed node objects returned by WorkerClass.get_current_node().
Each node type exposes type-specific methods.
"""

from __future__ import annotations
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from netcrawl.client import ApiClient


class Edge:
    """Represents a connection from the current node to another."""
    def __init__(self, id: str, other_node: str):
        self.id = id
        self.other_node = other_node

    def __repr__(self):
        return f"Edge(id={self.id!r}, other_node={self.other_node!r})"


class BaseNode:
    """Base class for all node types. Provides common properties."""

    def __init__(self, info: dict, client: 'ApiClient', worker_id: str):
        self._info = info
        self._client = client
        self._worker_id = worker_id
        self.id: str = info['id']
        self.type: str = info['type']
        self.label: str = info.get('label', '')
        self.data: dict = info.get('data', {})
        self.edges: list[Edge] = [Edge(e['id'], e['otherNode']) for e in info.get('edges', [])]

    @property
    def is_unlocked(self) -> bool:
        return bool(self.data.get('unlocked'))

    @property
    def is_infected(self) -> bool:
        return bool(self.data.get('infected'))

    @property
    def upgrade_level(self) -> int:
        return self.data.get('upgradeLevel', 0)

    def __repr__(self):
        return f"{self.__class__.__name__}(id={self.id!r}, label={self.label!r})"


class HubNode(BaseNode):
    """The central hub node."""
    pass


class ResourceNode(BaseNode):
    """A resource-producing node (data)."""

    @property
    def resource_type(self) -> str:
        return self.data.get('resource', '')

    @property
    def rate(self) -> int:
        return self.data.get('rate', 0)

    @property
    def is_mineable(self) -> bool:
        return bool(self.data.get('mineable'))

    def mine(self) -> dict:
        """Mine this resource node. Requires a pickaxe equipped on the worker.
        Creates a drop on the ground that can be collected.

        Returns: { ok, drop: { type: 'data_fragment', amount } }
        """
        result = self._client.action("compute_mine", {})
        # Use the existing mine action via pickaxe
        # Note: mining is done through the pickaxe equipment, not the node directly
        raise NotImplementedError(
            "Mining is done through equipment: self.pickaxe.mine(). "
            "The ResourceNode provides info about the node."
        )


class RelayNode(BaseNode):
    """A network relay node."""
    pass


class ComputeNode(BaseNode):
    """A compute puzzle node. Workers can request puzzles and submit answers."""

    @property
    def difficulty(self) -> str:
        return self.data.get('difficulty', 'easy')

    @property
    def reward_resource(self) -> str:
        return self.data.get('rewardResource', 'data')

    @property
    def solve_count(self) -> int:
        return self.data.get('solveCount', 0)

    def get_task(self) -> 'ComputeTask':
        """Request a puzzle from this compute node.

        Returns a ComputeTask with task_id, parameters, hint, and difficulty.
        Raises ValueError if on cooldown or not at a compute node.

        Example:
            node = self.get_current_node()
            task = node.get_task()
            answer = task.parameters['a'] + task.parameters['b']
            result = node.submit(task.task_id, answer)
        """
        result = self._client.action("compute", {})
        if not result.get("ok"):
            raise ValueError(f"get_task() failed: {result.get('error')}")
        return ComputeTask(
            task_id=result['taskId'],
            parameters=result['params'],
            hint=result.get('hint', ''),
            difficulty=result.get('difficulty', 'easy'),
        )

    def submit(self, task_id: str, answer: Any) -> dict:
        """Submit an answer to a compute puzzle.

        Returns: { ok, correct: bool, reward?: { type, amount } }
        If incorrect: { ok, correct: False, expected: ..., got: ... }
        """
        return self._client.action("submit", {"taskId": task_id, "answer": answer})


class ComputeTask:
    """A puzzle task from a ComputeNode."""

    def __init__(self, task_id: str, parameters: dict, hint: str, difficulty: str):
        self.task_id = task_id
        self.parameters = parameters
        self.hint = hint
        self.difficulty = difficulty

    def __repr__(self):
        return f"ComputeTask(id={self.task_id!r}, hint={self.hint!r})"


class APINode(BaseNode):
    """
    An API endpoint node. Receives external requests that workers must handle.

    Usage:
        node = self.get_current_node()
        request = node.poll_for_request()
        if request is None:
            return  # no pending requests

        # SECURITY: always check has_token!
        if not request.has_token:
            self.warn("Dropping unauthenticated request")
            return

        # Process the request
        a = request.body['a']
        b = request.body['b']
        node.respond(request.id, {'result': a + b})
    """

    def poll_for_request(self) -> 'APIRequestObj | None':
        """Pick up the next pending request, or None if queue is empty."""
        result = self._client.action("api_poll", {})
        if not result.get("ok"):
            return None
        req = result.get("request")
        if req is None:
            return None
        return APIRequestObj(
            id=req['id'],
            type=req['type'],
            body=req['body'],
            has_token=req['has_token'],
            deadline_tick=req['deadline_tick'],
            reward=req['reward'],
        )

    def respond(self, request_id: str, response_data: dict) -> dict:
        """
        Submit a response to a request.

        Returns: { ok, correct, credits_earned, speed_bonus }
        On error: { ok: False, error, reason }

        WARNING: responding to a request with has_token=False
        will cause a SECURITY BREACH and infect this node!
        """
        return self._client.action("api_respond", {
            "requestId": request_id,
            "responseData": response_data,
        })

    def get_stats(self) -> dict:
        """Get queue statistics: { pending, completed, failed, expired }"""
        return self._client.action("api_stats", {})


class APIRequestObj:
    """A request received at an API Node."""

    def __init__(self, id: str, type: str, body: dict, has_token: bool, deadline_tick: int, reward: dict):
        self.id = id
        self.type = type
        self.body = body
        self.has_token = has_token
        self.deadline_tick = deadline_tick
        self.reward = reward

    def __repr__(self):
        token_str = "AUTH" if self.has_token else "NO-AUTH"
        return f"APIRequest(id={self.id!r}, type={self.type!r}, [{token_str}])"


class CacheNodeType(BaseNode):
    """A cache structure node. Workers access it via get_service()."""
    pass


class EmptyNode(BaseNode):
    """An empty buildable slot."""
    pass


class LockedNode(BaseNode):
    """An unknown locked node."""
    pass


class InfectedNode(BaseNode):
    """An infected node."""
    pass


# ── Factory ──────────────────────────────────────────────────────────────────

NODE_TYPE_MAP = {
    'hub': HubNode,
    'resource': ResourceNode,
    'relay': RelayNode,
    'compute': ComputeNode,
    'api': APINode,
    'cache': CacheNodeType,
    'empty': EmptyNode,
    'locked': LockedNode,
    'infected': InfectedNode,
}


def create_node(info: dict, client: 'ApiClient', worker_id: str) -> BaseNode:
    """Factory: create the appropriate typed node from server info."""
    node_type = info.get('type', '')
    cls = NODE_TYPE_MAP.get(node_type, BaseNode)
    return cls(info, client, worker_id)

File: packages/sdk-js/src/nodes.ts
/**
 * nodes.ts
 *
 * Typed node objects returned by WorkerClass.getCurrentNode().
 * Each node type exposes type-specific methods.
 */

import type { ApiClient } from './client.js';

export class NodeEdge {
  readonly id: string;
  readonly otherNode: string;

  constructor(id: string, otherNode: string) {
    this.id = id;
    this.otherNode = otherNode;
  }

  toString(): string {
    return `NodeEdge(id=${this.id}, otherNode=${this.otherNode})`;
  }
}

export class BaseNode {
  protected _info: Record<string, unknown>;
  protected _client: ApiClient;
  protected _workerId: string;
  readonly id: string;
  readonly type: string;
  readonly label: string;
  readonly data: Record<string, unknown>;
  readonly edges: NodeEdge[];

  constructor(info: Record<string, unknown>, client: ApiClient, workerId: string) {
    this._info = info;
    this._client = client;
    this._workerId = workerId;
    this.id = info['id'] as string;
    this.type = info['type'] as string;
    this.label = (info['label'] as string) ?? '';
    this.data = (info['data'] as Record<string, unknown>) ?? {};
    const rawEdges = (info['edges'] as Array<{ id: string; otherNode: string }>) ?? [];
    this.edges = rawEdges.map(e => new NodeEdge(e.id, e.otherNode));
  }

  get isUnlocked(): boolean {
    return Boolean(this.data['unlocked']);
  }

  get isInfected(): boolean {
    return Boolean(this.data['infected']);
  }

  get upgradeLevel(): number {
    return (this.data['upgradeLevel'] as number) ?? 0;
  }

  toString(): string {
    return `${this.constructor.name}(id=${this.id}, label=${this.label})`;
  }
}

export class HubNode extends BaseNode {}

export class ResourceNode extends BaseNode {
  get resourceType(): string {
    return (this.data['resource'] as string) ?? '';
  }

  get rate(): number {
    return (this.data['rate'] as number) ?? 0;
  }

  get isMineable(): boolean {
    return Boolean(this.data['mineable']);
  }

  mine(): never {
    throw new Error(
      'Mining is done through equipment: self.pickaxe.mine(). ' +
      'The ResourceNode provides info about the node.'
    );
  }
}

export class ComputeNode extends BaseNode {
  get difficulty(): string {
    return (this.data['difficulty'] as string) ?? 'easy';
  }

  get rewardResource(): string {
    return (this.data['rewardResource'] as string) ?? 'data';
  }

  get solveCount(): number {
    return (this.data['solveCount'] as number) ?? 0;
  }

  async getTask(): Promise<ComputeTask> {
    const result = await this._client.action('compute', {});
    if (!result['ok']) {
      throw new Error(`getTask() failed: ${result['error']}`);
    }
    return new ComputeTask(
      result['taskId'] as string,
      result['params'] as Record<string, unknown>,
      (result['hint'] as string) ?? '',
      (result['difficulty'] as string) ?? 'easy',
    );
  }

  async submit(taskId: string, answer: unknown): Promise<Record<string, unknown>> {
    return this._client.action('submit', { taskId, answer });
  }
}

export class ComputeTask {
  readonly taskId: string;
  readonly parameters: Record<string, unknown>;
  readonly hint: string;
  readonly difficulty: string;

  constructor(taskId: string, parameters: Record<string, unknown>, hint: string, difficulty: string) {
    this.taskId = taskId;
    this.parameters = parameters;
    this.hint = hint;
    this.difficulty = difficulty;
  }

  toString(): string {
    return `ComputeTask(id=${this.taskId}, hint=${this.hint})`;
  }
}

export class APINode extends BaseNode {
  async pollForRequest(): Promise<APIRequestObj | null> {
    const result = await this._client.action('api_poll', {});
    if (!result['ok']) {
      return null;
    }
    const req = result['request'] as Record<string, unknown> | null;
    if (req == null) {
      return null;
    }
    return new APIRequestObj(
      req['id'] as string,
      req['type'] as string,
      (req['body'] as Record<string, unknown>) ?? {},
      Boolean(req['has_token']),
      (req['deadline_tick'] as number) ?? 0,
      (req['reward'] as Record<string, unknown>) ?? {},
    );
  }

  async respond(requestId: string, responseData: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this._client.action('api_respond', {
      requestId,
      responseData,
    });
  }

  async getStats(): Promise<Record<string, unknown>> {
    return this._client.action('api_stats', {});
  }
}

export class APIRequestObj {
  readonly id: string;
  readonly type: string;
  readonly body: Record<string, unknown>;
  readonly hasToken: boolean;
  readonly deadlineTick: number;
  readonly reward: Record<string, unknown>;

  constructor(
    id: string,
    type: string,
    body: Record<string, unknown>,
    hasToken: boolean,
    deadlineTick: number,
    reward: Record<string, unknown>,
  ) {
    this.id = id;
    this.type = type;
    this.body = body;
    this.hasToken = hasToken;
    this.deadlineTick = deadlineTick;
    this.reward = reward;
  }

  toString(): string {
    const tokenStr = this.hasToken ? 'AUTH' : 'NO-AUTH';
    return `APIRequest(id=${this.id}, type=${this.type}, [${tokenStr}])`;
  }
}

export class CacheNodeType extends BaseNode {}
export class EmptyNode extends BaseNode {}
export class LockedNode extends BaseNode {}
export class InfectedNode extends BaseNode {}

// Factory

const NODE_TYPE_MAP: Record<string, new (info: Record<string, unknown>, client: ApiClient, workerId: string) => BaseNode> = {
  hub: HubNode,
  resource: ResourceNode,
  compute: ComputeNode,
  api: APINode,
  cache: CacheNodeType,
  empty: EmptyNode,
  locked: LockedNode,
  infected: InfectedNode,
};

export function createNode(info: Record<string, unknown>, client: ApiClient, workerId: string): BaseNode {
  const nodeType = (info['type'] as string) ?? '';
  const Cls = NODE_TYPE_MAP[nodeType] ?? BaseNode;
  return new Cls(info, client, workerId);
}
